walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,114 @@
+    List<TimeRange> unavailableTimeWithOverlap = getAttendeesUnavailableTime(events, attendees);
+    List<TimeRange> unavailableTime = getNoOverlapTime(unavailableTimeWithOverlap);
+    List<TimeRange> availableTime = getAvailableTime(unavailableTime, request.getDuration());
+    return availableTime;
+  }
+
+  /** Get a collection of all attendees' unavailable time ranges (due to other events). */
+  private List<TimeRange> getAttendeesUnavailableTime(
+      Collection<Event> events, Collection<String> meetingAttendees) {
+    List<TimeRange> unavailableTime = new ArrayList<>();
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      // If any attendee in the meeting joins that event as well, the time range of the event is
+      // added to the unavailableTime collection.
+      if (!Collections.disjoint(meetingAttendees, eventAttendees)) {
+        unavailableTime.add(event.getWhen());
+      }
+    }
+    return unavailableTime;
+  }
+
+  /** Return a collection of time ranges in the time order without any overlap among each other. */
+  private List<TimeRange> getNoOverlapTime(List<TimeRange> unavailableTime) {
+    List<TimeRange> noOverlapTime = new ArrayList<>();
+    for (TimeRange newTime : unavailableTime) {
+      if (noOverlapTime.isEmpty()) {
+        noOverlapTime.add(newTime);
+        continue;
+      }
+      // Compare and combine the new time with any existing time range in collection in order.
+      List<TimeRange> localNoOverlapTime = new ArrayList<>();
+      for (int i = 0; i < noOverlapTime.size(); i++) {
+        TimeRange existingTime = noOverlapTime.get(i);
+        // Case 1. Existing time is earlier than new time
+        if (existingTime.end() <= newTime.start()) {
+          localNoOverlapTime.add(existingTime);
+        } else if (existingTime.start() >= newTime.end()) {
+          // Case 2. New time is earlier than existing time
+          localNoOverlapTime.add(newTime);
+          localNoOverlapTime.addAll(noOverlapTime.subList(i, noOverlapTime.size()));
+          break;
+        } else if (existingTime.contains(newTime)) {
+          // Case 3. Existing time contains new time
+          localNoOverlapTime.addAll(noOverlapTime.subList(i, noOverlapTime.size()));
+          break;
+        } else if (newTime.contains(existingTime)) {
+          // Case 4. New time contains existing time
+        } else {
+          // Case 5. Overlap (no contain) exists between existing time and new time
+          newTime = TimeRange.fromStartEnd(Math.min(newTime.start(), existingTime.start()),
+              Math.max(newTime.end(), existingTime.end()), false);
+        }
+      }
+      // Check whether newTime has never been included because it is the last one.
+      if (localNoOverlapTime.isEmpty()) {
+        localNoOverlapTime.add(newTime);
+      } else {
+        if (localNoOverlapTime.get(localNoOverlapTime.size() - 1).end() < newTime.end()) {
```

Comment:
Can this condition return false?

Reply:
Yes. 

For example, when the last checked "new time" is 9:00-10:00, but there exists some time later than that for events (e.g. a 20:00-21:00) that has been checked earlier. "localNoOverlapTime.get(localNoOverlapTime.size() - 1).end()" will return 21:00, because all time range in localNoOverlapTime is ordered, while the "new time" is picked up from the initial unsorted list and "newTime.end()" returns 10:00.
In this case, 21 is not smaller than 10, the condition returns false, so the new time does not need to be added.

In the other case, the end of the last time slot in the sorted "localNoOverlapTime" can be 11:00. While "newTime" can be 20:00-21:00, but it never has chance to be added earlier, then "new time" 20:00-21:00 needs to be added here.